### PR TITLE
Add encrypted_p2p_keys and generator command

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

File: core/store/migrations/migration1574659987/migrate.go [hunk 1]
```diff
@@ -2,15 +2,16 @@ package migration1574659987
...
-	}
-	return nil
+	return db.Exec(`
+		CREATE TABLE encrypted_secret_keys (
+			public_key character varying(68) PRIMARY KEY,
+			vrf_key text NOT NULL,
+			created_at timestamp with time zone NOT NULL,
+			updated_at timestamp with time zone NOT NULL
+		);
+	`).Error
```
Comment (topliceanu) [inline core/store/migrations/migration1574659987/migrate.go:16]: Why did you change this older migration?

File: core/store/migrations/migration1596021087/migrate.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package migration1596021087
+
+import (
+	"github.com/jinzhu/gorm"
+)
+
+// Migrate creates the encrypted_p2p_keys table and renames encryped_secret_keys to the more specific encrypted_vrf_keys
+func Migrate(tx *gorm.DB) error {
+	return tx.Exec(`
+		ALTER TABLE encrypted_secret_keys RENAME TO encrypted_vrf_keys;
```
Comment (topliceanu) [inline core/store/migrations/migration1596021087/migrate.go:10]: This could be split into two migrations. 
The reason is you're doing two separate things here: renaming encrypted_secret_keys to be more specific and adding a new encrypted_p2p_keys table. 
I would go so far as to suggest you split this PR in 2!

File: core/store/models/p2pkey/p2p_key.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+// type is added to the beginning of the passwords for
+// P2P key, so that the keys can't accidentally be mis-used
+// in the wrong place
+func adulteratedPassword(auth string) string {
+	s := "p2pkey" + auth
+	return s
+}
+
+func (k Key) Encrypt(auth string, p ...ScryptParams) (s EncryptedP2PKey, err error) {
```
Comment (topliceanu) [inline core/store/models/p2pkey/p2p_key.go:65]: I don't understand why this method needs to have this signature? 
I can't find any call of Encrypt with params other than the password?
Is there an interface that `Key` needs to implement? If so, can you add a comment on this method with that interface's name.

File: go.mod [hunk 4]
```diff
@@ -10,7 +10,7 @@ require (
 	github.com/bitly/go-simplejson v0.5.0
 	github.com/bmizerany/assert v0.0.0-20160611221934-b7ed37b82869 // indirect
 	github.com/boj/redistore v0.0.0-20160128113310-fc113767cd6b // indirect
-	github.com/btcsuite/btcd v0.0.0-20190115013929-ed77733ec07d
+	github.com/btcsuite/btcd v0.20.1-beta
```
Comment (topliceanu) [inline go.mod:13]: Nice! but should really be in a different PR, it's not related to this change!

File: core/store/models/p2pkey/p2p_key.go (hunk 3 above)
Comment (kaleofduty) [inline core/store/models/p2pkey/p2p_key.go:65]: Could we make `DefaultScryptParams` exported instead of the variadic-as-optional-arg thing?

File: core/cmd/app.go [hunk 5]
```diff
@@ -259,6 +259,20 @@ func NewApp(client *Client) *cli.App {
...
+				cli.Command{
+					Name:   "p2p",
+					Usage:  "Local commands for administering the node's p2p layer",
+					Hidden: !client.Config.Dev(),
+					Subcommands: cli.Commands{
+						{
+							Name: "create",
+							Usage: format(`Create a p2p key, encrypted with password from the
+               password file, and store it in the database.`),
+							Flags:  flags("password, p"),
```
Comment (topliceanu) [inline core/cmd/app.go:271]: The name of the flag `password` does not suggest it's a file path. To me it suggests it's the password in clear! Can you rename to `password-file` and update the Usage message accordingly. 
Also, `-p` seems to be a required field. Can you make that specific in the Usage message!

File: core/store/models/p2pkey/p2p_key.go (hunk 3 above)
Comment (kaleofduty) [inline core/store/models/p2pkey/p2p_key.go:65]: `Encrypt()` is a bit confusing on a key (you would expect it to encrypt some data potentially), how about `Encrypted()` instead

File: go.mod (hunk 4 above)
Comment (kaleofduty) [inline go.mod:13]: what's up with the beta version?

File: core/store/migrations/migration1574659987/migrate.go (hunk 1 above)
Comment (samsondav) [reply core/store/migrations/migration1574659987/migrate.go:16]: Because it was using gorm's horrible automigrate feature which changes when you update the struct. I changed it to be what it WOULD have always done.

File: core/store/migrations/migration1596021087/migrate.go (hunk 2 above)
Comment (samsondav) [reply core/store/migrations/migration1596021087/migrate.go:10]: Why?

File: go.mod (hunk 4 above)
Comment (samsondav) [reply go.mod:13]: go mod did this by itself

File: core/cmd/app.go (hunk 5 above)
Comment (samsondav) [reply core/cmd/app.go:271]: I did it this way so it was the same as VRF generation

File: go.mod (hunk 4 above)
Comment (samsondav) [reply go.mod:13]: go did this by itself

Comment (topliceanu) [reply go.mod:13]: I was also suspicious about this so I checked the published versions:
```
    v0.20.1-beta – Nov 13, 2019
    v0.13.0-beta2 – Dec 19, 2018
    v0.13.0-beta – Nov 18, 2018
```
Source: https://pkg.go.dev/github.com/btcsuite/btcd?tab=versions

🤦

File: core/store/models/p2pkey/p2p_key.go (hunk 3 above)
Comment (samsondav) [reply core/store/models/p2pkey/p2p_key.go:65]: Not really sure what you mean by this. I got rid of the variadic args since it wasn't used, does it look ok now?

File: go.mod (hunk 4 above)
Comment (topliceanu) [reply go.mod:13]: I want to clarify: this PR adds `libp2p/go-libp2p-core`, which depends on `btcsuite/btcd`:
```
➜  chainlink git:(feature/#174061930-libp2p-keygen) ✗ go mod graph  | grep btcsuite | grep libp2p
github.com/libp2p/go-libp2p-core@v0.6.0 github.com/btcsuite/btcd@v0.20.1-beta
```
`go mod graph` prints the adjacency list of the full dependency graph

As you’ve seen above, the latest version of btcd has a beta tag. 
Also, go will resolve dependencies when a command that compiles the source is executed, that’s `go run`, `go build`, `go install`, `go test`, and maybe others. Sam did a `go test`! 

I thinks there is no error being introduced here, @secure-lorenz  and I were just surprised by the `beta` tag. 

This blog post gives more details about go modules:  https://blog.golang.org/using-go-modules. There is a paragraph in there that explains what is considered the "latest" version:

```
(“Latest” is defined as the latest tagged stable (non-prerelease) version, or else the latest tagged prerelease version, or else the latest untagged version.)
```

Comment (topliceanu) [reply go.mod:13]: I want to clarify: this PR adds `libp2p/go-libp2p-core`, which depends on `btcsuite/btcd`:
```
➜  chainlink git:(feature/#174061930-libp2p-keygen) ✗ go mod graph  | grep btcsuite | grep libp2p
github.com/libp2p/go-libp2p-core@v0.6.0 github.com/btcsuite/btcd@v0.20.1-beta
```
`go mod graph` prints the adjacency list of the full dependency graph

As you’ve seen above, the latest version of btcd has a beta tag. 
Also, go will resolve dependencies when a command that compiles the source is executed, that’s `go run`, `go build`, `go install`, `go test`, and maybe others. Sam did a `go test`! 

I thinks there is no error being introduced here and I were just surprised by the `beta` tag. 

This blog post gives more details about go modules:  https://blog.golang.org/using-go-modules. There is a paragraph in there that explains what is considered the "latest" version:

> (“Latest” is defined as the latest tagged stable (non-[prerelease](https://semver.org/#spec-item-9)) version, or else the latest tagged prerelease version, or else the latest untagged version.)

File: core/services/vrf/vrf_solidity_crosscheck_test.go [hunk 6]
```diff
@@ -359,8 +359,8 @@ func TestVRF_MarshalProof(t *testing.T) {
 		sk := randomScalar(t, r)
 		skNum := secp256k1.ToInt(sk)
 		nonce := randomScalar(t, r)
-		seed := randomUint256(t, r)
-		proof, err := vrf.GenerateProofWithNonce(skNum, seed, secp256k1.ToInt(nonce))
+		sd := randomUint256(t, r)
```
Comment (kaleofduty) [inline core/services/vrf/vrf_solidity_crosscheck_test.go:362]: why rename this var?

Comment (samsondav) [reply core/services/vrf/vrf_solidity_crosscheck_test.go:362]: Go shadow complained about it.

I have to assume something got merged such that it is also broken in develop.
